### PR TITLE
Add AI image generation commands

### PR DESCRIPTION
**Pull request by glcraft**
Add two commands that generate images from sdxl turbo (request sent to hugging face)
* `/sdxl_turbo` : Generate image and send the picture in the command interaction
* `/gen_emoji` : Generate temporary emojis from sdxl turbo request

### CODE REVIEW

File: src/commands/ai/sdxl_turbo.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        ),
+    async execute(interaction) {
+        await interaction.deferReply();
+        const replyError = async (msgError: string): Promise<void> => {
+            await interaction.editReply(msgError);
+        };
+
+        // Discord slash command parameters
+        const prompt = interaction.options.get('prompt')?.value as string;
+        const seed = (interaction.options.get('seed')?.value as number) || Math.floor(Math.random() * 12013012031030);
```
Comment (Arthur-Devs) [inline src/commands/ai/sdxl_turbo.ts]: Peux être mettre les constantes par défaut en haut du fichier 
et au moins on pourra mettre dynamiquement dans la description des options

File: src/commands/ai/gen_emoji.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+/*
+ * @command     - gen_emoji
+ * @description - Génère des emojis avec SDLXL Turbo! (temporaire, 1 journée)
+ * @permission  - None
+ */
+export const GEN_EMOJI: Command = {
+    data: new SlashCommandBuilder()
+        .setName('gen_emoji')
+        .setDescription('Génère des emojis avec SDLXL Turbo! (temporaire, 1 journée)')
```
Comment (Arthur-Devs) [inline src/commands/ai/gen_emoji.ts]: Peux être utiliser la constante DURATION_TMP_EMOJI ici pour mettre la durée au moins on pourra changer la variable si on veut

**Review by Arthur-Devs — DISMISSED** [review]

File: src/events/cleanup_emojis.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            const guild = await oauth2guild.fetch();
+            const emojis = await guild.emojis.fetch();
+            for await (const emoji of emojis.values()) {
+                if (!emoji.name?.startsWith('tmp_')) {
+                    continue;
+                }
+                const timestamp = /tmp_(\d+)/.exec(emoji.name)?.[1];
+                if (!timestamp) {
+                    continue;
+                }
```
Comment (vic1707) [inline src/events/cleanup_emojis.ts]: the regex can do the startWith by itself right ?

File: src/events/cleanup_emojis.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+/*
+ * @event       - onReady - Cleanup temporary emojis
+ * @description - Gestion de la suppression des emojis temporaires
+ * @permission  - None
+ */
+export const CLEANUP_EMOJIS: BotEvent = {
+    name: 'Clean up temporary emojis',
+    listenTo: 'ready',
+    async execute(client) {
+        // Gestion de la suppression des emojis temporaires
```
Comment (vic1707) [inline src/events/cleanup_emojis.ts]: rm, is already in JSdoc

File: src/events/cleanup_emojis.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            const emojis = await guild.emojis.fetch();
+            for await (const emoji of emojis.values()) {
+                if (!emoji.name?.startsWith('tmp_')) {
+                    continue;
+                }
+                const timestamp = /tmp_(\d+)/.exec(emoji.name)?.[1];
+                if (!timestamp) {
+                    continue;
+                }
+                const diff = Math.max(0, DURATION_TMP_EMOJI - (Date.now() - Number(timestamp)));
```
Comment (vic1707) [inline src/events/cleanup_emojis.ts]:
```suggestion
                const diff = Math.max(0, DURATION_TMP_EMOJI - Date.now() - timestamp);
```

File: src/events/cleanup_emojis.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                setTimeout(() => void emoji.delete(), diff);
+            }
+        }
+    },
+    once: true,
```
Comment (vic1707) [inline src/events/cleanup_emojis.ts]: can you move that up so its clearer when reading the event code

+ personnal note: add that prop to JSDoc

File: src/commands/ai/sdxl_turbo.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import type { Command } from '@commands';
+import { SlashCommandBuilder } from 'discord.js';
+
+import { EventReader } from './hugging_face.ts';
+
+const DEFAULT_VALUE = {
+    strength: 0.7,
+    steps: 2,
+    get seed(): number {
+        return Math.floor(Math.random() * 12013012031030);
```
Comment (vic1707) [inline src/commands/ai/sdxl_turbo.ts]: what's that ?

File: src/commands/ai/sdxl_turbo.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            option.setName('steps').setDescription("Le nombre d'étapes (2 par défaut)").setRequired(false),
+        )
+        .addNumberOption((option) =>
+            option.setName('seed').setDescription('La graine (aléatoire par défaut)').setRequired(false),
+        ),
+    async execute(interaction) {
+        await interaction.deferReply();
+        const replyError = async (msgError: string): Promise<void> => {
+            await interaction.editReply(msgError);
+        };
```
Comment (vic1707) [inline src/commands/ai/sdxl_turbo.ts]: global function/utility class ?

File: src/commands/ai/sdxl_turbo.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            prompt,
+            strength,
+            steps,
+            seed,
+        });
+        if (!image) {
+            return replyError('Un problème est survenu...');
+        }
+        await interaction.editReply({
+            content: interaction.options.get('seed') === null ? `Graine: ${seed}` : null,
```
Comment (vic1707) [inline src/commands/ai/sdxl_turbo.ts]: would be nice to see the input string

**Review by vic1707 — CHANGES_REQUESTED** [review]

File: src/commands/ai/sdxl_turbo.ts (hunk 8 above)
Comment (glcraft) [reply src/commands/ai/sdxl_turbo.ts:37]: *magic number* from the api. made a const on the top.

File: src/commands/ai/sdxl_turbo.ts (hunk 9 above)
Comment (glcraft) [reply src/commands/ai/sdxl_turbo.ts]: Done

File: src/events/cleanup_emojis.ts (hunk 5 above)
Comment (glcraft) [reply src/events/cleanup_emojis.ts]: timestamp is a string, not a number. It doesn't work

File: src/commands/ai/sdxl_turbo.ts (hunk 8 above)
Comment (glcraft) [reply src/commands/ai/sdxl_turbo.ts]: `interaction` is a captured variable.
But it can now be replaced by a `throw 'error'`
What do you think ?

Comment (vic1707) [reply src/commands/ai/sdxl_turbo.ts]: Either pass the interaction or yeah throw (will need to show the right error message in the interactionCreate event)

File: src/events/cleanup_emojis.ts (hunk 5 above)
Comment (vic1707) [reply src/events/cleanup_emojis.ts]: Screenshot_2024-02-20-17-26-07-97_e4424258c8b8649f6e67d283a50a2cbc
Welcome to JS

Comment (glcraft) [reply src/events/cleanup_emojis.ts]: lol, you forget about eslint. he cries like a bitch

Comment (vic1707) [reply src/events/cleanup_emojis.ts]: You can try 
`Math.max(0, DURATION_TMP_EMOJI - Date.now() - +timestamp);`
Which should work

File: src/commands/ai/gen_emoji.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        .addStringOption((option) => option.setName('prompt').setDescription('Le prompt').setRequired(true))
+        .addNumberOption((option) =>
+            option.setName('seed').setDescription('La graine (aléatoire par défaut)').setRequired(false),
+        ),
+    async execute(interaction) {
+        await interaction.deferReply();
+        const hf_options = {
+            ...DEFAULT_VALUE,
+            prompt: interaction.options.get('prompt', true).value as string,
+            seed: (interaction.options.get('seed')?.value as number) ?? Math.floor(Math.random() * 12013012031030),
```
Comment (vic1707) [inline src/commands/ai/gen_emoji.ts]: default value ?

File: src/commands/ai/hugging_face.ts [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+    const html = await res.text();
+
+    const reg = new RegExp(`(${BASE_URL.replace('/', '\\/')}\\/--replicas\\/\\w+)`).exec(html);
+    if (!reg?.[1]) {
+        return null;
+    }
+    return reg[1];
+}
+
+let ROOT: Option<string> = await getRoot();
```
Comment (vic1707) [inline src/commands/ai/hugging_face.ts]: move to the top ?

File: src/commands/ai/hugging_face.ts [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+
+async function getFileFromRoot(path: string, force: boolean = true): Promise<Option<ArrayBuffer>> {
+    if (!ROOT) {
+        await getRoot();
```
Comment (vic1707) [inline src/commands/ai/hugging_face.ts]: no assignment

File: src/commands/ai/hugging_face.ts [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+        });
+
+        await event_reader.process();
+
+        return event_reader.image();
+    }
+    private async processEvent(evt: Event): Promise<boolean> {
+        switch (evt.msg) {
+            // case "estimation":
+            //     break;
```
Comment (vic1707) [inline src/commands/ai/hugging_face.ts]: put a default case that breaks ?

File: src/commands/ai/hugging_face.ts [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+                        this.data.input.strength,
+                        this.data.input.steps,
+                        this.data.input.seed,
+                    ]))
+                ) {
+                    return false;
+                }
+                break;
+            // case "process_starts":
+            //     break;
```
Comment (vic1707) [inline src/commands/ai/hugging_face.ts]: put a default case that breaks ?

File: src/commands/ai/hugging_face.ts [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+
+        return event_reader.image();
+    }
+    private async processEvent(evt: Event): Promise<boolean> {
+        switch (evt.msg) {
+            // case "estimation":
+            //     break;
+            case 'send_data':
+                if (
+                    !(await send_data(evt.event_id, this.data.session_hash, [
```
Comment (vic1707) [inline src/commands/ai/hugging_face.ts]: :vomit:

File: src/commands/ai/hugging_face.ts [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+                        this.data.input.seed,
+                    ]))
+                ) {
+                    return false;
+                }
+                break;
+            // case "process_starts":
+            //     break;
+            case 'process_completed':
+                if (evt.success) {
```
Comment (vic1707) [inline src/commands/ai/hugging_face.ts]: negate for beautiful guards :pray:

File: src/commands/ai/hugging_face.ts [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+
+export const MAX_SEED_API = 12013012031030;
+
+export const DEFAULT_VALUE = {
+    strength: 0.7,
+    steps: 2,
+    get seed(): number {
+        return Math.floor(Math.random() * MAX_SEED_API);
+    },
+};
```
Comment (vic1707) [inline src/commands/ai/hugging_face.ts]: move to the top

File: src/utils/duration.ts [hunk 18]
```diff
@@ -0,0 +1,58 @@
+const seconds = (sec: number): number => sec * 1000;
+const minutes = (min: number): number => seconds(min * 60);
+const hours = (hrs: number): number => minutes(hrs * 60);
+const days = (da: number): number => hours(da * 24);
```
Comment (vic1707) [inline src/utils/duration.ts]: secToMs
minToMS 
etc... ?
maybe those directly inside Duration ?

File: src/commands/ai/hugging_face.ts (hunk 15 above)
Comment (glcraft) [reply src/commands/ai/hugging_face.ts]: :swallow_vomit: 😊

**Review by vic1707 — DISMISSED** [review]
maybe move the hugging face module to a separate folder and subdivise it into something like
```
src
| - utils
     |- hugging face
         |- types.ts
         |- index.ts
         |- request handler
```
and add it to tsconfig as something like `@utils/huggin-face` 🤔 

up to you

**Review by vic1707 — APPROVED** [review]

**Pull request merged**